src/main/java/de/cronn/reflection/util/RecordSupport.java

```diff
@@ -0,0 +1,289 @@
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.jetbrains.annotations.Nullable;
+import org.objenesis.ObjenesisHelper;
+
+import net.bytebuddy.ByteBuddy;
+
+class RecordSupport {
+
+	private static volatile Boolean hasIsRecordMethod = null;
+
+	private static final ClassValue<Class<?>> dummySubclasses = new ClassValue<Class<?>>() {
+		@Override
+		protected Class<?> computeValue(Class<?> type) {
+			return new ByteBuddy()
+				.subclass(type)
+				.make()
+				.load(RecordSupport.class.getClassLoader())
+				.getLoaded();
+		}
+	};
+
+	static boolean isRecord(Class<?> beanClass) {
+		if (hasIsRecordMethod == null) {
+			hasIsRecordMethod = computeHasIsRecordMethod();
+		}
+		if (!hasIsRecordMethod) {
+			return false;
+		}
+		return invokeMethod(beanClass, "isRecord");
+	}
+
+	private static boolean computeHasIsRecordMethod() {
+		try {
+			Class.class.getMethod("isRecord");
+			return true;
+		} catch (NoSuchMethodException ignore) {
+			return false;
+		}
+	}
+
+	static boolean isRecord(Object bean) {
+		return isRecord(bean.getClass());
+	}
+
+	static class RecordComponentInfo {
+		private final String name;
+		private final Class<?> type;
+		private final Method accessor;
+
+		RecordComponentInfo(String name, Class<?> type, Method accessor) {
+			this.name = name;
+			this.type = type;
+			this.accessor = accessor;
+		}
+
+		public String getName() {
+			return name;
+		}
+
+		Class<?> getType() {
+			return type;
+		}
+
+		Method getAccessor() {
+			return accessor;
+		}
+	}
+
+	static <T> Method findMethod(Class<T> beanClass, TypedPropertyGetter<T, ?> propertyGetter) {
+		RecordComponentInfo[] recordComponents = getRecordComponents(beanClass);
+		Object[] uniqueValues = buildUniqueValues(recordComponents);
+		try {
+			Constructor<T> recordConstructor = getRecordConstructor(beanClass);
+			T record = ClassUtils.createInstance(recordConstructor, uniqueValues);
+
+			Object value = propertyGetter.get(record);
+
+			if (needsFallbackToComponentSearch(uniqueValues, value)) {
+				return exhaustiveComponentSearch(value, beanClass, propertyGetter, uniqueValues, recordConstructor);
+			}
+
+			int componentIndex = ArrayUtils.indexOf(uniqueValues, value);
+			if (componentIndex < 0) {
+				throw new IllegalArgumentException("Failed to find a method for the given property getter");
+			}
+			return recordComponents[componentIndex].getAccessor();
+		} catch (ReflectiveOperationException e) {
+			throw new ReflectionRuntimeException(e);
+		}
+	}
+
+	private static RecordComponentInfo[] getRecordComponents(Class<?> beanClass) {
+		Object[] recordComponents = invokeMethod(beanClass, "getRecordComponents");
```

Comment:
Not sure if this a problem for you to use reflection here. But it's Friday and maybe idea is just stupid, but would like to discuss it :)

Create new project responsible for handling Records. Compiled in java 17. Add it to reflection-util project as compile time dependency and optional in runtime.
Use classes from new project only when Record.class is found. When records are missing class won't be loaded, no issues. When record exists user has to add new library.

Actually it might be more complex for users. But would it work?

Reply:
I guess it could work but I think the developer experience for the library users would not be so nice.

The reflective method invocation has a rather low overhead, once the JVM is warmed up so I guess I would like to keep this approach for now.

Reply:
I guess it could work but I think the developer experience for the library users would not be so nice.

The reflective method invocation has a rather low overhead, once the JVM is warmed up so I guess I would like to keep this approach for now.